src/master.py

```diff
@@ -81,22 +89,135 @@ def aws_update_projects(pname):
     "calls state.highstate on ALL projects matching <projectname>-*"
     return aws_update_many_projects([pname])
 
+@cached
+def cached_master_ip(stackname):
+    return core.stack_data(stackname)[0]['private_ip_address']
+
 @debugtask
 @requires_aws_stack
-def remaster_minion(stackname, master_ip=None):
+def remaster(stackname, new_master_stackname):
     "tell minion who their new master is. deletes any existing master key on minion"
+    # TODO: turn this into a decorator
+    import cfn
+    # start the machine if it's stopped
+    # you might also want to acquire a lock so alfred doesn't stop things
+    cfn._check_want_to_be_running(stackname, 1)
 
-    if not master_ip:
-        newest_master = last(core.find_master_servers(core.active_aws_stacks(core.find_region())))
-        if not newest_master:
-            raise core.NoMasterException("no master servers found")
-        master_ip = core.stack_data(newest_master)[0]['private_ip_address']
-
+    master_ip = cached_master_ip(new_master_stackname)
     print('re-mastering %s to %s' % (stackname, master_ip))
 
+    context = context_handler.load_context(stackname)
+    if context.get('ec2') == True:
+        # TODO: duplicates bad ec2 data wrangling in cfngen.build_context
+        # ec2 == True for some reason, which is completely useless
+        LOG.warn("bad context for stack: %s", stackname)
+        context['ec2'] = {}
+        context['project']['aws']['ec2'] = {}
+
+    if context['ec2'].get('master_ip') == master_ip:
+        LOG.info("already remastered: %s", stackname)
+        try:
+            utils.get_input('any key to skip, ctrl-c to carry on')
+            return
+        except KeyboardInterrupt:
+            LOG.info("not skipping")
+
+    LOG.info("setting new master address")
+    context = cfngen.set_master_address(context, master_ip) # mutator
```

Comment:
if this is a mutator it shouldn't really return anything, just set stuff in `context`?